src/clean-core/range_ref.hh

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include <initializer_list>
+
+#include <clean-core/function_ref.hh>
+#include <clean-core/is_range.hh>
+#include <clean-core/iterator.hh>
+
+namespace cc
+{
+/// a type-erased range of elements that can be converted to T
+template <class T>
+struct range_ref
+{
+    /// empty range
+    range_ref()
+    {
+        _for_each = [](cc::function_ref<void(T)>) {};
+    }
+
+    /// NOTE: make_range_ref is easier to use
+    template <class F, cc::enable_if<std::is_invocable_r_v<void, F, cc::function_ref<void(T)>>> = true>
+    range_ref(F&& for_each_fun)
+    {
+        _for_each = for_each_fun;
+    }
```

Comment:
Do we need SFINAE here, or would a static_assert do?

Reply:
unconstrained ctor templates "hijack" copy/move ctors

Reply:
Ah i see